### DISCUSSION TITLE
Proof of concept language client

### DISCUSSION THREAD

**Discussion by alcarney** [openlawlibrary/pygls#408] [Show and tell]
[Screencast_20231031_233015-1.webm](https://github.com/openlawlibrary/pygls/assets/2675694/391f9628-4717-4a99-a8e8-11e3cdefb8f1)

Over in the [lsp-devtools](https://github.com/swyddfa/lsp-devtools) repo, I now have a proof of concept language client working! It's built using textual and the `BaseLanguageClient` we have in `pygls`. It currently only supports basic document sync and completion requests  - but it's a start!

Rather than trying to be a "real editor", I hope that this can grow into nice way to interact with a server while visualizing the traffic sent between it and the client

**Comment by tombh** [comment]
That's amazing 😲 For so many reasons! Have you built a text editor!? Are the tree viewer, text editor and completion modal part of `textual`? And the log of requests and the payload explorer are just wow, _chef's kiss_.

**Reply by alcarney** [reply]
All the UI elements you see are built into textual, all I've done is connect them up to our language client :)  

> Have you built a text editor? 

I wouldn't call it a text editor yet, I haven't even implemented saving files! 😅

**Reply by tombh** [reply]
Oh okay, still impressive!

So what ideas do you have for it? It could be embedded into a web page with Pyodide? And can those requests and payloads panes be streamed as raw JSON? Then they could live as split panes in a normal editor like Vim or VSCode.

**Reply by alcarney** [reply]
> So what ideas do you have for it?

I quite like the idea of giving the client multiple "personalities" where it can try its best to pretend to be VSCode, or pretend to be Vim by presenting the server with a different set of client capabilities. It would be nice not to have to setup 10 different editors just to see how your server adapts to them. 

It would also be nice to try and visualize as much of the data coming back from the server as possible e.g. previewing the result of selecting a completion, without actually having to select it. - However, implementing all that will be a non-trivial task, so no idea if it would ever materialize! :sweat_smile: 

> Then they could live as split panes in a normal editor like Vim or VSCode.

You can do that already! 

![Screenshot depicting the `lsp-devtools inspect` command being used side-by-side with Neovim](https://user-images.githubusercontent.com/2675694/212438877-d332dd84-14b4-4568-b36f-4c3e04d4f95f.png)

The devtools component is available as a standalone app via the `lsp-devtools inspect` command (well, `lsp-devtools tui` on the current PyPi release)

> And can those requests and payloads panes be streamed as raw JSON?

Yes! There's the `lsp-devtools record`  command for that :smile: - I actually have [documentation written](https://lsp-devtools.readthedocs.io/en/latest/lsp-devtools/guide/record-command.html) for this one if you're interested

In order to get access to the LSP messages though, both `lsp-devtools inspect` and `lsp-devtools record` need your client to invoke the language server via the [`lsp-devtools agent`](https://lsp-devtools.readthedocs.io/en/latest/lsp-devtools/guide/getting-started.html) command

>  It could be embedded into a web page with Pyodide?

Potentially? maybe, perhaps? I know that [textual-web](https://textual.textualize.io/blog/2023/09/06/what-is-textual-web/) now exists, but don't know enough about how it works to know if it can be used for that....

**Comment by tombh** [comment]
> I quite like the idea of giving the client multiple "personalities" 

Oh, so that's what the Textual editor is for, acting as a chameleon editor?

> You can do that already! ... `lsp-devtools inspect`

You think of everything!

>  I actually have documentation written for this one if you're interested

And of course the documentation is top tier 🧐

> Potentially? maybe, perhaps? I know that textual-web now exists, but don't know enough about how it works to know if it can be used for that...

I can't help but think of a LSP Playground, where people can share and even interact with small in-editor LSP examples, all whilst seeing debug feedback. Maybe a bit ambitious.

**Comment by tombh** [comment]
Have you seen https://github.com/microsoft/monaco-editor?

I noticed it had this in its README about adding extensions:

>❓ I've written an extension for VS Code, will it work on the Monaco Editor in a browser?
>No.
>    _Note: If the extension is fully based on the LSP and if the language server is authored in JavaScript, then it would be possible._

Just been thinking about embedding your idea in a webpage.

**Reply by alcarney** [reply]
> Have you seen https://github.com/microsoft/monaco-editor?

Yep, that was what was powered an old attempt I had at doing something similar  (https://github.com/openlawlibrary/pygls/pull/291, this [commit](https://github.com/openlawlibrary/pygls/pull/291/commits/166afdf8387fd7074af6ffadf62d6002caab3527) in particular), though that never got as far as showing the captured messages.

We could always try to resurrect it?  :)

**Reply by tombh** [reply]
Oh! I hadn't connected the dots. Of course, you were already working on this. So do you think we could have a LSP-compliant in-browser editor, with live LSP debugging? And then some simple backend that lets us share things like playground.pygls.org/wK3xAa (it doesn't actually exist, just an example) that might reproduce some bug or showoff some feature!

**Reply by alcarney** [reply]
It's definitely possible, whether that's quick and/or easy to implement is another matter!

Off the top of my head there's 3 approaches we could explore.
- The option with (probably) the least amount of effort would be to
   - Add WASI support to `pygls`
   - Expand the VSCode Playground extension to include the message visualisation stuff (or even just enable the verbose logging!) and publish it as a web extension
   - Piggyback off of https://vscode.dev and use github repos as the storage backend
- Look into building something with [JupyterLite](https://jupyterlite.readthedocs.io/en/stable/), though I'm not sure how easy it is to hook a wasm language server into [jupyterlab-lsp](https://github.com/jupyter-lsp/jupyterlab-lsp)
- Turn #291 into a full blown custom web app, it would offer the most flexibility but would also require the most work

**Reply by tombh** [reply]
So one of the differences between approach 1 and 3 is that 1 uses the full VSCode editor and 3 just uses the Monaco editor?

I guess that the, "if the language server is authored in JavaScript" part of that Monaco FAQ answer ("If the extension is fully based on the LSP and if the language server is authored in JavaScript, then it would be possible.") doesn't mean that a Pygls LSP in Pyodide would count as "in JS"? 😬

**Reply by alcarney** [reply]
> So one of the differences between approach 1 and 3 is that 1 uses the full VSCode editor and 3 just uses the Monaco editor?
 
Yes.

I think the FAQ is trying to say that for your LSP server to work with Monaco you need to be able to run the server within the browser itself - which typically means writing the server in JS. However, thanks to WASM, we can write our servers with pygls and just use JS to [wrap the WASM program](https://github.com/alcarney/pygls/blob/166afdf8387fd7074af6ffadf62d6002caab3527/examples/playground/src/serverWorker.ts#L64)

The same limitations would exist with vscode.dev - though it can [handle more of the plumbing](https://github.com/microsoft/vscode-wasm) for us

**Reply by tombh** [reply]
I think just because I have don't have so much experience with VSCode, and that I can picture the Monaco solution in my head, I imagined that the Monaco approach is easier. But it sounds like piggybacking off of vscode.dev will just be more plug and play.
